File: src/main/java/leetcode/hot100/LeetCode416.java
package leetcode.hot100;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/5 12:07
 */
public class LeetCode416 {

    //https://www.youtube.com/watch?v=z_VLFGzQQtk

    //和零钱兑换类似 322

    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int n : nums) {
            sum += n;
        }
        if (sum % 2 == 1) {
            return false;
        }
        int target = sum / 2;
        boolean[] dp = new boolean[target + 1];
        for (int num : nums) {
            // 为什么从后往前遍历
            // 硬币只能用一次，从左往右遍历，会重复拿硬币，后面的结果会受到前面的影响
            // 从右向左遍历，越减越小，左边的不会受右边的影响
            // https://xth8013.com/website/blogArticle/detail/111
            for (int currSum = target; currSum > 0; currSum--) {
                if (currSum - num >= 0) {
                    dp[currSum] = dp[currSum] || dp[currSum - num];
                }
            }
        }
        return dp[target];
    }
}

File: src/main/go/leetcode/lists/hot100/LeetCode416.go
package hot100

func canPartition(nums []int) bool {
	if len(nums) == 0 {
		return true
	}
	sum := 0
	for i := 0; i < len(nums); i++ {
		sum += nums[i]
	}
	if sum%2 != 0 {
		return false
	}
	target := sum / 2
	dp := make([]bool, target+1)
	dp[0] = true
	for _, nu := range nums {
		for i := target; i > 0; i-- {
			if i-nu >= 0 {
				dp[i] = dp[i] || dp[i-nu]
			}
		}
	}
	return dp[target]
}
